net/device/IotHub.cs

```diff
@@ -0,0 +1,157 @@
+
+        private const string _contentEncoding = "utf-8";
+        private const string _contentType = "application/json";
+
+        private volatile bool _isConnected;
+        private volatile bool _wasEverConnected;
+        private volatile ConnectionStatus _connectionStatus;
+        private volatile DeviceClient _deviceClient;
+
+        private static readonly JsonSerializerOptions _jsonSerializerOptions = new JsonSerializerOptions { IgnoreNullValues = true };
+
+        public IDictionary<string, string> IotProperties { get; } = new Dictionary<string, string>();
+
+        public IotHub(Logger logger, string deviceConnectionString, TransportType transportType)
+        {
+            _logger = logger;
+            _deviceConnectionString = deviceConnectionString;
+            _transportType = transportType;
+            _deviceClient = null;
+        }
+
+        public async Task InitializeAsync()
+        {
+            _lifetimeControl.Wait();
+
+            Cleanup();
+
+            _deviceClient = DeviceClient.CreateFromConnectionString(_deviceConnectionString, _transportType);
+            _deviceClient.SetConnectionStatusChangesHandler(ConnectionStatusChangesHandler);
+            await _deviceClient.OpenAsync().ConfigureAwait(false);
+
+            _lifetimeControl.Release();
+        }
+
+        public void Cleanup(bool force = false)
+        {
+            if (_deviceClient != null
+                && _wasEverConnected
+                && (force || _connectionStatus == ConnectionStatus.Disconnected))
+            {
+                _deviceClient.Dispose();
+                _deviceClient = null;
+                _wasEverConnected = false;
+            }
+        }
+
+        public async Task SendTelemetryAsync(object telemetryObject, IDictionary<string, string> extraProperties = null)
+        {
+            Debug.Assert(telemetryObject != null);
+
+            string message = JsonSerializer.Serialize(telemetryObject, _jsonSerializerOptions);
+            Debug.Assert(!string.IsNullOrWhiteSpace(message));
+
+            var iotMessage = new Message(Encoding.UTF8.GetBytes(message))
+            {
+                ContentEncoding = _contentEncoding,
+                ContentType = _contentType,
+            };
+
+            foreach (var prop in IotProperties)
+            {
+                iotMessage.Properties.TryAdd(prop.Key, prop.Value);
+            }
+
+            if (extraProperties != null
+                && extraProperties.Any())
+            {
+                foreach (var prop in extraProperties)
+                {
+                    iotMessage.Properties.TryAdd(prop.Key, prop.Value);
+                }
+            }
+            await _deviceClient.SendEventAsync(iotMessage).ConfigureAwait(false);
+            _logger.Trace($"Sent message [{message}]");
+        }
+
+        private void ConnectionStatusChangesHandler(ConnectionStatus status, ConnectionStatusChangeReason reason)
```

Comment:
Based off Abhipsa's implementation in the reconnect sample. I haven't yet added the exception handling logic, nor have I made use of the _isConnected variable (pretty sure I want to prevent telemetry from being sent when retrying, but not sure what that looks like yet).